src/aggregation/metric/top_hits.rs

```diff
@@ -0,0 +1,350 @@
+use std::collections::{BinaryHeap, HashMap};
+use std::fmt::Formatter;
+
+use serde::ser::SerializeMap;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+use super::{TopHitsMetricResult, TopHitsVecEntry};
+use crate::aggregation::bucket::Order;
+use crate::aggregation::intermediate_agg_result::{
+    IntermediateAggregationResult, IntermediateMetricResult,
+};
+use crate::aggregation::segment_agg_result::SegmentAggregationCollector;
+use crate::{DocAddress, SegmentOrdinal};
+
+/// # Top Hits
+///
+/// The top hits aggregation is a useful tool to answer questions like:
+/// - "What are the most recent posts by each author?"
+/// - "What are the most popular items in each category?"
+///
+/// It does so by keeping track of the most relevant document being aggregated,
+/// in terms of a sort criterion that can consist of multiple fields and their
+/// sort-orders (ascending or descending).
+///
+/// `top_hits` should not be used as a top-level aggregation. It is intended to be
+/// used as a sub-aggregation, inside a `terms` aggregation or a `filters` aggregation,
+/// for example.
+///
+/// The following example demonstrates a request for the top_hits aggregation:
+/// ```JSON
+/// {
+///     "aggs": {
+///         "top_authors": {
+///             "terms": {
+///                 "field": "author",
+///                 "size": 5
+///             }
+///         },
+///         "aggs": {
+///             "top_hits": {
+///                 "size": 2,
+///                 "from": 0
+///                 "sort": [
+///                     { "date": "desc" }
+///                 ]
+///             }
+///         }
+/// }
+/// ```
+///
+/// This request will return an object containing the top two documents, sorted
+/// by the `date` field in descending order. You can also sort by multiple fields, which
+/// helps to resolve ties. The aggregation object for each bucket will look like:
+/// ```JSON
+/// {
+///     "hits": [
+///         {
+///           "id": <doc_address>,
+///           score: [<time_u64>]
+///         },
+///         {
+///             "id": <doc_address>,
+///             score: [<time_u64>]
+///         }
+///     ]
+/// }
+/// ```
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
+pub struct TopHitsAggregation {
+    sort: Vec<KeyOrder>,
+    size: usize,
+    from: Option<usize>,
+}
+
+#[derive(Debug, Clone, PartialEq, Default)]
+struct KeyOrder(String, Order);
```

Comment:
can't we use Derive here?

Reply:
Is there built-in derive to map `(f, o)` => `{"f": o}` or `{field: k, order: o}` => `{"f": "o}`?

Reply:
Ah I see, custom serialization is probably the best way to do this.

Can you change it into a struct with field names?

Reply:
Looks much better now [7e507be](https://github.com/quickwit-oss/tantivy/pull/2198/commits/7e507bea6b7a9a7cb3a6de7c24e74aed727443c4)